File: 2019-kakao-blind-recruitment/Q1.실패율.py
from collections import Counter


def compose(*functions):
    return reduce(lambda f, g: lambda x: f(g(x)), functions, lambda x: x)


def to_failure(fail_user, challenge_user):
    return fail_user / challenge_user if challenge_user != 0 else 0


def solution(N, stages):
    game_result = [[i, 0, 0] for i in range(N + 1)]
    stage_counter = Counter(stages)

    for stage, count in stage_counter.items():
        for i in range(1, stage):
            game_result[i][2] += count

        if stage <= N:
            game_result[stage][1] += count
            game_result[stage][2] += count

    game_result = [[result[0], to_failure(*result[1:])]
                   for result in game_result[1:]]

    game_result.sort(key=lambda x: -x[1])

    return [result[0] for result in game_result]

File: 2019-kakao-blind-recruitment/Q1.실패율.js
function go(...args) {
	return args.reduce((x, f) => f(x));
}

function map(f, iter) {
	const res = [];

	for (const el of iter) {
		res.push(f(el));
	}
	return res;
}

function toFailure([stage, ...result]) {
	return [stage, result[0] / result[1]];
}

function counter(arr) {
	const counterMap = new Map();

	arr.forEach(el => {
		if (!counterMap.has(el)) counterMap.set(el, 0);
		counterMap.set(el, counterMap.get(el) + 1);
	});

	return Array.from(counterMap);
}

function solution(N, stages) {
	const gameResults = Array.from(Array(N + 1), (_, idx) => [idx, 0, 0]);
	const stageCounter = counter(stages);

	for (const [stage, count] of stageCounter) {
		for (let i = 1; i <= Math.min(stage, N); i++) {
			if (i === stage) {
				gameResults[stage][1] += count;
			}

			gameResults[i][2] += count;
		}
	}

	return go(
		gameResults.slice(1),
		gameResults => map(toFailure, gameResults),
		failureResults => failureResults.sort((a, b) => b[1] - a[1]),
		failureResults => map(failureResult => failureResult[0], failureResults),
	);
}
